From 198f7818c95261bf3fd14110df1004a4e7c35c9a Mon Sep 17 00:00:00 2001
From: Adrian Smith <triode1@btinternet.com>
Date: Sun, 20 May 2012 00:17:48 +0100
Subject: [PATCH 7/7] ignore tuned buffer values when using plug layer

snd_pcm_close can crash when switching if pcm is opened with large buffer time values when using plug layer
avoid this by hard coding touch defult buffer values for use with low samplerate streams requring sample rate conversion
---
 src/audio/decode/decode_alsa_backend.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/audio/decode/decode_alsa_backend.c b/src/audio/decode/decode_alsa_backend.c
index 0ec3a9951..162aa381d 100644
--- a/src/audio/decode/decode_alsa_backend.c
+++ b/src/audio/decode/decode_alsa_backend.c
@@ -628,6 +628,7 @@ static int _pcm_open(struct decode_alsa *state,
 	snd_pcm_uframes_t size;
 	snd_ctl_elem_id_t *id;
 	snd_pcm_hw_params_t *hw_params;
+	int plug = 0;
 
 	hw_params = (snd_pcm_hw_params_t *) alloca(snd_pcm_hw_params_sizeof());
 
@@ -666,6 +667,7 @@ static int _pcm_open(struct decode_alsa *state,
 
 			LOG_INFO("Reopening device %s in plug mode as %s", device, plug_device);
 			device = plug_device;
+			plug = 1;
 
 			snd_pcm_close(*pcmp);
 			if ((err = snd_pcm_open(pcmp, device, mode, 0)) < 0) {
@@ -736,14 +738,14 @@ static int _pcm_open(struct decode_alsa *state,
 	}
 
 	/* set buffer and period times */
-	val = state->period_count;
+	val = !plug ? state->period_count : 2;  // safe value when plug layer used, else snd_pcm_close can crash
 	if ((err = snd_pcm_hw_params_set_periods_near(*pcmp, hw_params, &val, 0)) < 0) {
 		LOG_ERROR("Unable to set period size %s", snd_strerror(err));
 		return err;
 	}
 	state->period_count = val;
 
-	val = state->buffer_time;
+	val = !plug ? state->buffer_time : 20000; // safe value when plug layer used, else snd_pcm_close can crash
 	dir = 1;
 	if ((err = snd_pcm_hw_params_set_buffer_time_near(*pcmp, hw_params, &val, &dir)) < 0) {
 		LOG_ERROR("Unable to set  buffer time %s", snd_strerror(err));
-- 
2.36.1

